### DISCUSSION TITLE
custom theme use css variables

### DISCUSSION THREAD

**Discussion by fauzanashari** [saadeghi/daisyui#1811] [Q&A]
Hello, I want to create my own custom theme with values of the properties is using css variable.

tailwind.config.js:

```
daisyui: {
    themes: [
      {
        myTheme: {
          "--customPrimary": "deeppink",
          "primary": "var(--customPrimary)",
          ....otherProperties
        },
      }
    ]
}
```

The expectation from code above is value of primary is deeppink but the error is appear `[postcss] Unable to parse color from string: var(--customPrimary)`.

How to solve this problem? I store the value in css variable bcause I'll reuse it vars in css/scss files

**Comment by saadeghi** [comment]
It's not working this way because the color value you provide needs to be parsed so daisyUI can add Tailwind's opacity value to it. If we allow any color value (for example an unknown CSS variable) we won't be able to add opacity value. And the opacity value depends on if the color is being used in a background color utility class or text color or border color, etc. So it can't be a CSS variable without parsing.

You can use this method to add custom colors with CSS variable:  
https://github.com/saadeghi/daisyui/discussions/653

**Reply by asharifauzan** [reply]
actually what i trying to achive is define style variables which can be shared to my custom theme and sass files. Currently what i hv been implementing is like this code below


tailwind.config.js
```
// single source of truth  variables
const vars = {
  primary: "deeppink",
  secondary: "gray"
}

module.exports = {
   ......
  daisyui: {
    themes: [
      {
        myTheme: {
          //send to scss files 
          "--primary": vars.primary,
          "--secondary": vars.secondary,

          "primary": vars.primary,
          "secondary": vars.secondary,
        },
      },
    ],
  },
 ....
```

_vars.scss
```
$primary: var(--primary);
$secondary: var(--secondary);

// use the vars
.my-loading {
  ...
  background: $primary;
  ...
}
```

I don't know is this best practice or not, but those code solve my problem to share style to custom theme and scss file. Any suggestion? thank in advance

**Reply by saadeghi** [reply]
> I don't know is this best practice or not

Depends on your needs. If you just want some CSS variables, completely independent from Tailwind CSS utility classes, then yes, it works.  

However in daisyUI itself, we need more than that because we need Tailwind CSS `opacity-*` utility classes to work everywhere. We need to assign the raw value of HSL (or RGB, or another color function) to the CSS variable.  
For example, it should be `--my-variable: 338 83% 66%`  
So then we can send it to Tailwind CSS color function like `mycolor: 'hsl(var(--my-variable) / <alpha-value>)',`.  
With this, Tailwind generates all utility classes for `mycolor` (`bg-mycolor`, `text-mycolor`, etc) and it also supports `opacity-*` utlity classes for them depending on where the color is being used (background, border, etc.)

**Reply by asharifauzan** [reply]
thank you for the information sir.

Currently what I'm implementing is code like my previous answer, but it seems like windows has an issue when I move all variables into separated files and updating my vars the update is not changing immediately. So I moved to Linux or WSL since this problem doesn't occurs in this OS. Thank you.

**Reply by saadeghi** [reply]
> windows has an issue when I move all variables into separated files and updating my vars the update is not changing immediately

That's strange 🤔  
What if you `@import` your variables file into the other file?

**Reply by asharifauzan** [reply]
> > windows has an issue when I move all variables into separated files and updating my vars the update is not changing immediately
> 
> That's strange 🤔 What if you `@import` your variables file into the other file?

the issue still appears unless I define all variables in tailwind.config.js directly as my code above

**Reply by saadeghi** [reply]
@asharifauzan Can you share a small reproduction repo?

**Reply by asharifauzan** [reply]
> @asharifauzan Can you share a small reproduction repo?

I created this repository https://github.com/asharifauzan/react-daisyui-theme or you can see the commit related to this instead https://github.com/asharifauzan/react-daisyui-theme/commit/f1ffdae8bdf303fb83db66450e734c0b117cbc98
